### DISCUSSION TITLE
Group Normalization/ optimization of current layernom and instance norm

### DISCUSSION THREAD

**Discussion by dhorka** [pyg-team/pytorch_geometric#2912] [Ideas]
I saw in the documentation that pytorch_geometric has the implementation of batch normalization,layer normalization, instance normalization but the group normalization is missing. Is there any plans to have the group normalization implemented in pytorch geometric?

**Comment by rusty1s** [comment]
The `GroupNorm` is indeed missing. Are you interested in contributing this feature?

**Reply by dhorka** [reply]
Is something that I will try to do it. But for the moment what I am trying to do is modify the layer_norm in order to be able to use Sparse Tensors. Because, yesterday after doing some tests I found that for my use case, just changing the batch norm for a layer norm the forward steps was 10x times slower... I was thinking that maybe doing this modification is going to help me to increase the speed of the network.

**Reply by rusty1s** [reply]
Note that `LayerNorm` does not make use of edge connectivity, so it cannot make use of the `SparseTensor` object (it just takes in `x` and `batch`). However, I agree that it is more computationally expensive than `BatchNorm` as it needs to normalize examples independently. I will look into it how to speed this up further. Any pointers appreciated.

**Reply by dhorka** [reply]
I haven't tried yet, but I was thinking to create the sparse tensor using the batch just to be able to use the sparse tensor. But I have to think more detail about it an check the current available operations with this sparse tensor.

**Reply by rusty1s** [reply]
Instead of using `SparseTensor`, you may be able to swap out the `scatter` calls with `segment_csr` calls from `torch-scatter`.

**Reply by dhorka** [reply]
I will check it this weekend, segment_csr should be faster than scatter operations?

**Reply by rusty1s** [reply]
Yes, it should be faster.

**Reply by dhorka** [reply]
I will check that during this week and I will come back to you

**Comment by dhorka** [comment]
I have tried as you suggested to use segment_csr. This is the implementation that I have done:

```
class LayerNorm(torch.nn.Module):
    def __init__(self, in_channels, eps=1e-5, affine=True):
        super(LayerNorm, self).__init__()

        self.in_channels = in_channels
        self.eps = eps

        if affine:
            self.weight = Parameter(torch.Tensor([in_channels]))
            self.bias = Parameter(torch.Tensor([in_channels]))
        else:
            self.register_parameter('weight', None)
            self.register_parameter('bias', None)

        self.reset_parameters()

    def reset_parameters(self):
        ones(self.weight)
        zeros(self.bias)

    def forward(self, x: Tensor, batch: OptTensor = None) -> Tensor:
        if batch is None:
            x = x - x.mean()
            out = x / (x.std(unbiased=False) + self.eps)

        else:
            batch_size = int(batch.max()) + 1

            norm = degree(batch, batch_size, dtype=x.dtype).clamp_(min=1)
            norm = norm.mul_(x.size(-1)).view(-1, 1)
            
            # sorting indx
            indx = torch.argsort(batch, dim=0, descending=False)
            batch = batch[indx]
            x = x[indx]
            
            # creating ptr
            u , count = torch.unique(batch, return_counts=True)
            batch_ptr = [torch.tensor(0, device=batch.device)]
            for i, c in enumerate(count):
                batch_ptr.append(batch_ptr[i] + c)
            batch_ptr = torch.stack(batch_ptr).to(batch.device)
            
            
            mean = segment_csr(x, batch_ptr,
                           reduce='add').sum(dim=-1, keepdim=True) / norm

            x = x - mean[batch]
            
            var = segment_csr(x * x, batch_ptr,
                            reduce='add').sum(dim=-1, keepdim=True)
            
            var = var / norm

            out = x / (var + self.eps).sqrt()[batch]

        if self.weight is not None and self.bias is not None:
            out = out * self.weight + self.bias

        return out

    def __repr__(self):
        return f'{self.__class__.__name__}({self.in_channels})'
```

I have checked the speed up on my use case, and I didn't notice to much speed up. I think that maybe the problem is the time that I am saving with the segment_csr is spended during the creation of the ptr. Do you think that it could be done more efficient this part?

Thanks!

**Reply by rusty1s** [reply]
You can simply access `data.ptr` instead of `data.batch`. There's no need to manually create it by yourself.

**Reply by dhorka** [reply]
I have created by myself the ptr for the cases where the normalization is applied inside of a graph conv after the expansion of the nodes, as it is the case of the edge_conv. Do you think that in this case it is not really necessary?

**Reply by rusty1s** [reply]
I'm not sure I understand. Can you give me an example?

**Reply by dhorka** [reply]
Yes, for instance this EdgeConv that I am defining below the nn is applied in the message passing step where the graph is expanded to be able to compute the edge messages. Then the original batch/ptr provided by the data structure is not going to work, isn't it?
```
edge_conv = EdgeConv(in_channels,
                    out_channels,
                    nn=torch_geometric.nn.Sequential('x, batch', [
                    (torch.nn.Linear(in_channels * 2, out_channels), 'x -> x'),
                    (LayerNorm(out_channels), 'x, batch -> x'),
                    (torch.nn.ReLU(inplace=True), 'x ->x'),
                    (torch.nn.Linear(out_channels, out_channels), 'x -> x'),
                    (LayerNorm(out_channels), 'x, batch -> x'),
                    (torch.nn.ReLU(inplace=True), 'x ->x'),
                    ]))
```

**Reply by rusty1s** [reply]
Instead of `batch`, you should be able to pass in `ptr`.

**Reply by dhorka** [reply]
It does not gonna work, isn't it? I mean when you compute x_i what you do is apply for instance x[edge_indx[0]] then in order to have the same shape for the batch you should compute the batch[edge_indx[0]]. That works for the batch vector, but the ptr is not that easy, isn'it? If I am not wrong the ptr contains the indx value where it should be applied the scatter operation, this indx values are going to change after obtaining the x_i. For this reason I do not believe that I am should be able to apply directly the ptr, in fact, in order to be able to apply the layernom on the edgeconv I had to create a custom version of it where I expand the batch as I explained before.

**Reply by rusty1s** [reply]
Ah, I see. Sorry for misunderstanding you. The problem is that you are operating on edge-level tensors, while `ptr` is node-level.

As a workaround, you can compute `ptr` from `batch` quite fast via:
```python
torch.bucketize(torch.arange(batch_size, device=batch.device, batch)
```

**Reply by dhorka** [reply]
I assume that in this scenario there is not way to avoid the sort that I am doing at the beggining, isn't? I will check about your proposal I didn't know about this torch method.

**Reply by rusty1s** [reply]
I don't think the sort is necessary, as `batch` is sorted by default.

**Reply by dhorka** [reply]
Thanks! I will check it and I will come back with some feedback.  For the moment what I had to do it is modify a little the torch.bucketize that you provided, just to obtain the same results that I was obtaining with my implementation:
```
batch_ptr = torch.cat([torch.tensor([0], device=batch.device),                                                                                    
                                    torch.bucketize(torch.arange(batch_size, device=batch.device), batch, right=True)])
```

**Reply by dhorka** [reply]
I have check it and there is an speedup but its still very slow in comparision to the batchnorm. This weekend I will try to prepare an example in colab to show a similar scenario to the one that I am using it. Also, I will try to think another way to do that, lets see if we can speedup more these kinds of normalizations.
